### ISSUE TITLE
bug: Signature Version 2 requests appear to break S3 persistence

### ISSUE THREAD

**Issue by mattkahl** [localstack/localstack#4297]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

S3 persistence works well until a Signature Version 2 command makes it into `recorded_api_calls.json`. When a Signature Version 2 command is in `recorded_api_calls.json`, "Waiting for all Localstack services to be ready" will poll seemingly endlessly.

Specifically, in my case, this is a `create-bucket` command sent from an S3 GUI, [Transmit](https://panic.com/transmit/).

The Signature Version 2 theory was formed after investigating the `recorded_api_calls.json` file and noticing the `AWS ` vs. `AWS4-HMAC-SHA256 ` prefix in the `Authorization` header.

### Expected Behavior

I expect any S3 API call that localstack handled successfully to be able to persist after the restarting of a container.

### How are you starting LocalStack?

With a `docker run` command

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)
    docker volume create my-localstack
    docker run --rm -it -p 4566:4566 -p 4571:4571 -e DEBUG=1 -e DATA_DIR=/mnt/data --mount 'type=volume,src=my-localstack,dst=/mnt/data' localstack/localstack:0.12.15

#### Step 1: What works
    awslocal s3api create-bucket --bucket test1
    awslocal s3api create-bucket --bucket test2
    # Restart via original `docker run` command persists correctly

#### Step 2: When it breaks
- Launch Transmit on macOS
- Connect to the localstack S3
- 'File -> New Bucket...'
- Create a new bucket named `test3`
- Restart via original `docker run` command hangs with endless "Waiting for all LocalStack services to be ready"

#### Alternative reproduction 
- Start localstack as described above (with the `my-localstack` volume)
- Copy the following recorded api calls
    ```
    {"a": "sqs", "m": "POST", "p": "/", "d": "QWN0aW9uPUxpc3RRdWV1ZXMmVmVyc2lvbj0yMDEyLTExLTA1", "h": {"Remote-Addr": "127.0.0.1", "Host": "localhost:4566", "Accept-Encoding": "identity", "Content-Type": "application/x-www-form-urlencoded; charset=utf-8", "User-Agent": "Boto3/1.17.106 Python/3.7.10 Linux/5.10.25-linuxkit Botocore/1.20.106", "X-Amz-Date": "20210713T203407Z", "Authorization": "AWS4-HMAC-SHA256 Credential=__internal_call__/20210713/us-east-1/sqs/aws4_request, SignedHeaders=content-type;host;x-amz-date, Signature=97141ce1eab936f2c267a99f24d084330f76e9d9f4aceb99dafdfb6dba7eb4e8", "Content-Length": "36", "X-Forwarded-For": "127.0.0.1, localhost:4566, 127.0.0.1, localhost:4566", "x-localstack-edge": "http://localhost:4566", "x-localstack-tgt-api": "sqs", "Connection": "close"}, "rd": "PExpc3RRdWV1ZXNSZXNwb25zZT48TGlzdFF1ZXVlc1Jlc3VsdD48L0xpc3RRdWV1ZXNSZXN1bHQ+PFJlc3BvbnNlTWV0YWRhdGE+PFJlcXVlc3RJZD4yWVA3VENTUUhTOENHOEFHR0QwSlRZODVBRlEzMDRMTlM5NzRNUUtITUdCSkxRVlBVREYyPC9SZXF1ZXN0SWQ+PC9SZXNwb25zZU1ldGFkYXRhPjwvTGlzdFF1ZXVlc1Jlc3BvbnNlPg=="}
    {"a": "s3", "m": "PUT", "p": "/test1", "d": "", "h": {"Remote-Addr": "172.17.0.1", "Host": "localhost:4566", "Accept-Encoding": "identity", "User-Agent": "aws-cli/2.0.46 Python/3.8.5 Darwin/20.3.0 source/x86_64 command/s3api.create-bucket", "X-Amz-Date": "20210713T203604Z", "X-Amz-Content-Sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", "Authorization": "AWS4-HMAC-SHA256 Credential=_not_needed_locally_/20210713/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=61ee7abf65a8d7c4b970e647c538d99094db15dbf5c13ae59dd2624946be1475", "Content-Length": "0", "X-Forwarded-For": "172.17.0.1, localhost:4566, 127.0.0.1, localhost:4566", "x-localstack-edge": "http://localhost:4566", "x-localstack-tgt-api": "s3", "content-type": "binary/octet-stream", "Connection": "close"}, "rd": "PENyZWF0ZUJ1Y2tldFJlc3BvbnNlIHhtbG5zPSJodHRwOi8vczMuYW1hem9uYXdzLmNvbS9kb2MvMjAwNi0wMy0wMSI+PENyZWF0ZUJ1Y2tldFJlc3BvbnNlPjxCdWNrZXQ+dGVzdDE8L0J1Y2tldD48L0NyZWF0ZUJ1Y2tldFJlc3BvbnNlPjwvQ3JlYXRlQnVja2V0UmVzcG9uc2U+"}
    {"a": "s3", "m": "PUT", "p": "/test2", "d": "", "h": {"Remote-Addr": "172.17.0.1", "Host": "localhost:4566", "Accept-Encoding": "identity", "User-Agent": "aws-cli/2.0.46 Python/3.8.5 Darwin/20.3.0 source/x86_64 command/s3api.create-bucket", "X-Amz-Date": "20210713T203930Z", "X-Amz-Content-Sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", "Authorization": "AWS4-HMAC-SHA256 Credential=_not_needed_locally_/20210713/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=10b9f7c867cf0afe77e18bdf299f45c9d2f73cd34d7dcab301c1f448e0d409c0", "Content-Length": "0", "X-Forwarded-For": "172.17.0.1, localhost:4566, 127.0.0.1, localhost:4566", "x-localstack-edge": "http://localhost:4566", "x-localstack-tgt-api": "s3", "content-type": "binary/octet-stream", "Connection": "close"}, "rd": "PENyZWF0ZUJ1Y2tldFJlc3BvbnNlIHhtbG5zPSJodHRwOi8vczMuYW1hem9uYXdzLmNvbS9kb2MvMjAwNi0wMy0wMSI+PENyZWF0ZUJ1Y2tldFJlc3BvbnNlPjxCdWNrZXQ+dGVzdDI8L0J1Y2tldD48L0NyZWF0ZUJ1Y2tldFJlc3BvbnNlPjwvQ3JlYXRlQnVja2V0UmVzcG9uc2U+"}
    {"a": "sqs", "m": "POST", "p": "/", "d": "QWN0aW9uPUxpc3RRdWV1ZXMmVmVyc2lvbj0yMDEyLTExLTA1", "h": {"Remote-Addr": "127.0.0.1", "Host": "localhost:4566", "Accept-Encoding": "identity", "Content-Type": "application/x-www-form-urlencoded; charset=utf-8", "User-Agent": "Boto3/1.17.106 Python/3.7.10 Linux/5.10.25-linuxkit Botocore/1.20.106", "X-Amz-Date": "20210713T204130Z", "Authorization": "AWS4-HMAC-SHA256 Credential=__internal_call__/20210713/us-east-1/sqs/aws4_request, SignedHeaders=content-type;host;x-amz-date, Signature=35b264b0b6185c157aa738faf05c1cf787fe99a2749bcdf522b270d0e3562ac1", "Content-Length": "36", "X-Forwarded-For": "127.0.0.1, localhost:4566, 127.0.0.1, localhost:4566", "x-localstack-edge": "http://localhost:4566", "x-localstack-tgt-api": "sqs", "Connection": "close"}, "rd": "PExpc3RRdWV1ZXNSZXNwb25zZT48TGlzdFF1ZXVlc1Jlc3VsdD48L0xpc3RRdWV1ZXNSZXN1bHQ+PFJlc3BvbnNlTWV0YWRhdGE+PFJlcXVlc3RJZD5aVkNLV0ZPVVpJV1ZOUEMzVzI3QkFERU9PVFlGNldZN0hEM1FFMENOQllURTBPVkM4SUNMPC9SZXF1ZXN0SWQ+PC9SZXNwb25zZU1ldGFkYXRhPjwvTGlzdFF1ZXVlc1Jlc3BvbnNlPg=="}
    {"a": "s3", "m": "PUT", "p": "/test3", "d": "", "h": {"Remote-Addr": "172.17.0.1", "Host": "localhost:4566", "Accept": "*/*", "Accept-Encoding": "gzip, deflate", "Accept-Language": "en-us", "Date": "Tue, 13 Jul 2021 20:42:48 GMT", "User-Agent": "Transmit/5.7.4", "Content-Length": "0", "Authorization": "AWS test:AgaZIyJLcDVk0Ye46dVJEOGyVOA=", "X-Forwarded-For": "172.17.0.1, localhost:4566, 127.0.0.1, localhost:4566", "x-localstack-edge": "https://localhost:4566", "x-localstack-tgt-api": "s3", "content-type": "binary/octet-stream", "Connection": "close"}, "rd": "PENyZWF0ZUJ1Y2tldFJlc3BvbnNlIHhtbG5zPSJodHRwOi8vczMuYW1hem9uYXdzLmNvbS9kb2MvMjAwNi0wMy0wMSI+PENyZWF0ZUJ1Y2tldFJlc3BvbnNlPjxCdWNrZXQ+dGVzdDM8L0J1Y2tldD48L0NyZWF0ZUJ1Y2tldFJlc3BvbnNlPjwvQ3JlYXRlQnVja2V0UmVzcG9uc2U+"}
    ```
- Launch a privileged container shell and paste the calls into `/var/lib/docker/volumes/my-localstack/_data/recorded_api_calls.json`
    ```
    docker run -it --privileged --pid=host debian nsenter -t 1 -m -u -n -i sh
    # Within shell
    vi /var/lib/docker/volumes/my-localstack/_data/recorded_api_calls.json
    # Paste the recorded api calls and close vi
    ```
- Restart via original `docker run` command hangs with endless "Waiting for all LocalStack services to be ready"

### Environment

```markdown
- OS: macOS 11.2.1
- LocalStack: 0.12.15
```

### Anything else?

I thought this was originally related to the #4033 but I now think it's the same problem @dmarkey was having in https://github.com/localstack/localstack/issues/4010#issuecomment-841647502.

Note the Authorization header difference between the SigV4 call via `awslocal s3 create-bucket`:
```
{"a": "s3", "m": "PUT", "p": "/test1", "d": "", "h": {"Remote-Addr": "172.17.0.1", "Host": "localhost:4566", "Accept-Encoding": "identity", "User-Agent": "aws-cli/2.0.46 Python/3.8.5 Darwin/20.3.0 source/x86_64 command/s3api.create-bucket", "X-Amz-Date": "20210713T203604Z", "X-Amz-Content-Sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", "Authorization": "AWS4-HMAC-SHA256 Credential=_not_needed_locally_/20210713/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=61ee7abf65a8d7c4b970e647c538d99094db15dbf5c13ae59dd2624946be1475", "Content-Length": "0", "X-Forwarded-For": "172.17.0.1, localhost:4566, 127.0.0.1, localhost:4566", "x-localstack-edge": "http://localhost:4566", "x-localstack-tgt-api": "s3", "content-type": "binary/octet-stream", "Connection": "close"}, "rd": "PENyZWF0ZUJ1Y2tldFJlc3BvbnNlIHhtbG5zPSJodHRwOi8vczMuYW1hem9uYXdzLmNvbS9kb2MvMjAwNi0wMy0wMSI+PENyZWF0ZUJ1Y2tldFJlc3BvbnNlPjxCdWNrZXQ+dGVzdDE8L0J1Y2tldD48L0NyZWF0ZUJ1Y2tldFJlc3BvbnNlPjwvQ3JlYXRlQnVja2V0UmVzcG9uc2U+"}
```
And the SigV2 call via Transmit:
```
{"a": "s3", "m": "PUT", "p": "/test3", "d": "", "h": {"Remote-Addr": "172.17.0.1", "Host": "localhost:4566", "Accept": "*/*", "Accept-Encoding": "gzip, deflate", "Accept-Language": "en-us", "Date": "Tue, 13 Jul 2021 20:42:48 GMT", "User-Agent": "Transmit/5.7.4", "Content-Length": "0", "Authorization": "AWS test:AgaZIyJLcDVk0Ye46dVJEOGyVOA=", "X-Forwarded-For": "172.17.0.1, localhost:4566, 127.0.0.1, localhost:4566", "x-localstack-edge": "https://localhost:4566", "x-localstack-tgt-api": "s3", "content-type": "binary/octet-stream", "Connection": "close"}, "rd": "PENyZWF0ZUJ1Y2tldFJlc3BvbnNlIHhtbG5zPSJodHRwOi8vczMuYW1hem9uYXdzLmNvbS9kb2MvMjAwNi0wMy0wMSI+PENyZWF0ZUJ1Y2tldFJlc3BvbnNlPjxCdWNrZXQ+dGVzdDM8L0J1Y2tldD48L0NyZWF0ZUJ1Y2tldFJlc3BvbnNlPjwvQ3JlYXRlQnVja2V0UmVzcG9uc2U+"}
```

**Comment by mopagemo** [comment]
I had the same problem on my setup and today I finally spent time to investigate it, coming to the same conclusion. I use https://metacpan.org/pod/Net::Amazon::S3 (perl module) to write to S3, and by default it uses V2 signatures, which prevented a successful startup. This line does not work with V2 signatures: https://github.com/localstack/localstack/blob/master/localstack/services/edge.py#L296

It throws an exception, which is then ignored (and not logged), which then leads to an unknown service being returned (`_unknown_`), which in turn makes it try to get a lock which is already obtained elsewhere (here: https://github.com/localstack/localstack/blob/master/localstack/services/edge.py#L161), thus being stuck endlessly waiting for the lock.

**Comment by whummer** [comment]
Thanks for reporting @mattkahl . This should be (hopefully) fixed in #4985 - can you please pull the `latest` Docker image (currently being built in our CI pipeline, available on Docker Hub in ~30mins) and give it another try? Please let us know if the problem persists.. Thanks!
